### ISSUE TITLE
Unable to read/write NFC tag

### ISSUE THREAD

**Issue by aaamoeder** [home-assistant/android#942]
**Home Assistant Android version:** beta-382-532ae9d-full

**Android version:** 10

**Phone model:** Oneplus7

**Home Assistant version:** 114.4

**Last working Home Assistant release (if known):** n/a

**Description of problem:** I cannot read or write an NFC tag which I can otherwise read or write fine using the "NFC tools" app.

**Traceback (if applicable):** n/a

```

```

**Screenshot of problem:**
![image](https://user-images.githubusercontent.com/20947118/93369365-938fc880-f84f-11ea-8f1e-1e19463f3ba5.png)
![image](https://user-images.githubusercontent.com/20947118/93369398-9be80380-f84f-11ea-8a9e-c8931885fafb.png)
![image](https://user-images.githubusercontent.com/20947118/93369410-a0acb780-f84f-11ea-84b8-eb045a2babd9.png)

**Additional information:** As you can see in the screenshot I have written a small text "by nfc tools" to the tag, so nothing should be wrong with the tag imo

**Comment by urfin78** [comment]
I have the same problem with a different kind of tag and I think is related to the size of the tag - mine is 46byte. 
If i try to write an uuid like: 554ee751-3adf-4f92-87f5-7517dcb60ec2 (which was created when i add a nfc tag manually in the homeassistant v0.115.0b11 web interface ) with the nfc tools app, it fails with an error stating not enough capacity.

**Comment by aaamoeder** [comment]
That might indeed be the case, 
I'm also unable to write any longer texts in the data field.
Perhaps they should give some minimum requirements for NFC tags before I order some that I can't use ..

**Comment by dshokouhi** [comment]
NTAG215 is what i have personally used...the only requirements I know of are mentioned here: https://companion.home-assistant.io/docs/integrations/universal-links

I am unsure of what size to use but I think if you stick to a re-writable tag you should be ok

**Comment by JBassett** [comment]
I think the minimum byte count is 70?  This is actually an example of the tag we write.
`https://www.home-assistant.io/tag/123e4567-e89b-12d3-a456-426614174000`

**Comment by hirschy81** [comment]
same problem running HA 115.2

**Comment by Armadill0** [comment]
Same problem here with 0.115.4 reading a NTAG216. Working fine with NFC Tools:
![photo_2020-09-28_21-26-33](https://user-images.githubusercontent.com/6536990/94476842-502f4580-01d1-11eb-9620-32652edcc6cd.jpg)

But when I try to write an NFC tag via Homoeassistant Android app:
![photo_2020-09-28_21-26-37](https://user-images.githubusercontent.com/6536990/94476850-545b6300-01d1-11eb-85ad-2eee8d47fd29.jpg)

I get this with my Poco F2 Pro:
![photo_2020-09-28_21-26-35](https://user-images.githubusercontent.com/6536990/94476858-59b8ad80-01d1-11eb-83ea-9ab507bad81d.jpg)

Size should be pretty sufficient with 888Bytes.
It's writeable with NFC Tools without issues.

I will order some NTAG215s to test them out anyway.

Any idea if the tag type really is the problem? Is this an error from the Home Assistant app or is this from the NFC Reader itself?

**Comment by Armadill0** [comment]
On the latest IOS app everything works like a charm btw with the exact same NFC tag.

**Comment by shmirgel** [comment]
I had similar issues. Reading/writing via NFC tools was OK, but had no success with Home Assistant android app.

Phone: Xiaomi Mi 8 pro
Android: 10
Home Assistant: version 2.5.0-full

The solution was to allow NFC for Home Assistant application(it was disabled for me): Touch and hold the HA application icon > App info > Other permissions > NFC > Accept

Then from HA application > Configuration > Tags > Add new tag with custom name(default Tag ID) and write it to the NFC tag.

**Comment by dshokouhi** [comment]
> The solution was to allow NFC for Home Assistant application(it was disabled for me): Touch and hold the HA application icon > App info > Other permissions > NFC > Accept

Can everyone else with this issue check this on their device?  My device does not have this option so NFC is always available.

**Comment by Armadill0** [comment]
> The solution was to allow NFC for Home Assistant application(it was disabled for me): Touch and hold the HA application icon > App info > Other permissions > NFC > Accept

@shmirgel Man that worked, thank you very much! :)

> > The solution was to allow NFC for Home Assistant application(it was disabled for me): Touch and hold the HA application icon > App info > Other permissions > NFC > Accept
> 
> Can everyone else with this issue check this on their device? My device does not have this option so NFC is always available.

@dshokouhi Checked, now works for me, sorry for bothering.

**Comment by dshokouhi** [comment]
@Armadill0 not a bother at all, its helpful for us to know this because we don't all use the same devices and they are not all built the same or function the same :)  I will leave this issue open for a few more days to let others respond.

**Comment by Armadill0** [comment]
@dshokouhi Great 😃 
As a sidenote: I was able to scan the tags without those rights. But writing was not possible. Maybe this is helpful for future issues.

**Comment by mr2c12** [comment]
I suffer from the very same problem, but unfortunately on my Huawei there is no such authorization for the app.
Whenever I try to read or write a tag, I get the "unable to write tag" error!

**Comment by Eddy805** [comment]
For future reference; I had a similar issue writing with my phone (OnePlus 5T) to a free promotional card. Reading went fine, but writing resulted in an error. However, after having written once with the NFC Tools, the HA app was able to write correctly -- multiple times and without failure.

**Comment by EagleAdam** [comment]
Hi,
I have a Samsung Galaxy S9, and am running 0.117.2. I can write tags without a problem but am getting the read error. In my permissions, I can't modify the permissions for NFC access but see that the Home Assistant app appears to have access:
![grafik](https://user-images.githubusercontent.com/44459480/98558261-84545680-22a5-11eb-9b01-75c97f189819.png)


I can write a tag without a problem:
![grafik](https://user-images.githubusercontent.com/44459480/98558289-8cac9180-22a5-11eb-96bb-9054c0304fe1.png)
![grafik](https://user-images.githubusercontent.com/44459480/98558302-90401880-22a5-11eb-86e4-1ac863427d36.png)

But then reading the same tag:
![grafik](https://user-images.githubusercontent.com/44459480/98558328-959d6300-22a5-11eb-80ce-64cab28aa8c8.png)
"Error while processing nfc tag"

What is strange is that the tag works fine on iOS... 
At the suggestion of dshokouhi on the discord server, I ran the android debug tool. I'm NOT a super advanced linux guy or debugger, so I had a hard time parsing through the logs. However I think I found the "moment" when the tag was scanned. Visibly it can read the card - as I see the full URL - but there is something preventing it from then properly opening it in home assistant. I'm happy to help debug by providing more logs, etc. See extract below (I removed the repeated '11-09 15:49:14.059 10321 10321 E TagReaderActivity:' from each line)

` Unable to handle url (nfc): https://www.home-assistant.io/tag/bb7fd21a-97ab-4453-9e8f-057672f647cc
 io.homeassistant.companion.android.common.data.integration.IntegrationException
 	at io.homeassistant.companion.android.common.data.integration.impl.IntegrationRepositoryImpl.scanTag(IntegrationRepositoryImpl.kt:233)
 	at io.homeassistant.companion.android.common.data.integration.impl.IntegrationRepositoryImpl$scanTag$1.invokeSuspend(Unknown Source:12)
 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
 	at android.os.Handler.handleCallback(Handler.java:883)
 	at android.os.Handler.dispatchMessage(Handler.java:100)
 	at android.os.Looper.loop(Looper.java:237)
 	at android.app.ActivityThread.main(ActivityThread.java:8167)
 	at java.lang.reflect.Method.invoke(Native Method)
 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)`

**Comment by dshokouhi** [comment]
@EagleAdam what version of the android app are you on? That error message occurs when we cant send that data back to HA.  I see you are using the HA frontend to read the tag, can you try it from the App Configuration page to see if it works there?

**Comment by EagleAdam** [comment]
Version 3.0.1-full. And unfortunately no, same error message in App Configuration.

**Comment by dshokouhi** [comment]
Are there any errors in the HA logs when this happens?

**Comment by EagleAdam** [comment]
No, afraid not...

**Comment by dshokouhi** [comment]
I think we may need to add some more debug logs around this to see where the issue is coming from.

**Comment by EagleAdam** [comment]
Happy to volunteer my phone / system if needed 😄 Just let me know what I need to do

**Comment by dshokouhi** [comment]
@EagleAdam can you try the full version debug APK from: https://github.com/home-assistant/android/actions/runs/354747509 it will install side by side the current app and you can change the device name to something else so you can remove it later on

This contains some additional ADB debug statements you can pull that should appear above the error you posted.

One will show the actual tag id `nfcTagId: <id>` and the other will start with `Unable to send tag data for` along with your HA URL (that you should remove from the log) and the data which should also contain the `nfcTagId` up above.  

You should at least see the first log containing the tag ID but I am not certain you will see the last log.  You may also see the last log more than once for each URL you have.

Comment back here with the log statements after you remove your URL please

**Comment by EagleAdam** [comment]
> 
> 
> @EagleAdam what version of the android app are you on? That error message occurs when we cant send that data back to HA. I see you are using the HA frontend to read the tag, can you try it from the App Configuration page to see if it works there?

Hi @dshokouhi and thanks for your help. So I don't see that specific text in the android debug logs. I see the following "NFC-related" entries (hard to parse them out from all the other stuff going on...)
```
NfcService: After 60 secs, send NfcConstants.SecMessage.MSG_TAG_CONNECTION
NfcService: [BIG DATA] TAG SUCCESS!!! PROTOCOL= 2
NfcService: tag value : NONE
NfcService: Handle launch intent with Knox Intent { act=android.nfc.action.NDEF_DISCOVERED dat=https://www.home-assistant.io/... pkg=io.homeassistant.companion.android (has extras) }
```

One note - in my earlier post I indicated that from the App Configuration I couldn't read the tag. In the normal version of the app I don't have a NFC Tags section in the App Configuration. However, in the debug version you shared, there is a NFC Tags section and I can successfully read the tag.

**Comment by dshokouhi** [comment]
> In the normal version of the app I don't have a NFC Tags section in the App Configuration. However, in the debug version you shared, there is a NFC Tags section and I can successfully read the tag.

sounds like you got something corrupted....in order to show the NFC section we check that the device reports it has the proper hardware and that HA version is at least 0.114.  I think you may need to clear data especially if the debug version looks correct since that part of the code did not change.  Follow these steps:

https://companion.home-assistant.io/docs/troubleshooting/faqs/#starting-fresh-with-the-android-app

**Comment by pepppek** [comment]
I have same problem in my system. On Iphone XS it works perfectly, but problem is with all android devices (Asus Zenphone 5, Android 10) and tablet SONY (SGP771, Android 7.1).  I can create NFC tag, but when you use it - shows message: "Error while processing nfc tag". I tried found NFC  in app setting, but nothing like this it wasnt here. Thanks for all advice

**Comment by tefracky** [comment]
Is there anything new on it? I am using a Galaxy S10+ with Android 11 and Home Assistant App 2021.2.2-full and cannot write NXP Mifare Classic 1k (ISO 1443-3A) NFC tags. Do you have any solution?

**Comment by dshokouhi** [comment]
> Is there anything new on it? I am using a Galaxy S10+ with Android 11 and Home Assistant App 2021.2.2-full and cannot write NXP Mifare Classic 1k (ISO 1443-3A) NFC tags. Do you have any solution?

did you try the solution 2 posts above?  If that fails then get us the crash logs, details can be found on the same page linked 2 posts above

**Comment by tefracky** [comment]
> > Is there anything new on it? I am using a Galaxy S10+ with Android 11 and Home Assistant App 2021.2.2-full and cannot write NXP Mifare Classic 1k (ISO 1443-3A) NFC tags. Do you have any solution?
> 
> did you try the solution 2 posts above? If that fails then get us the crash logs, details can be found on the same page linked 2 posts above

Yes, I tried. Here is the error:

Unable to write tag.
java.lang.Exception: Failed to format tag
	at io.homeassistant.companion.android.nfc.NFCUtil.writeMessageToTag(NFCUtil.kt:85)
	at io.homeassistant.companion.android.nfc.NFCUtil.createNFCMessage(NFCUtil.kt:25)
	at io.homeassistant.companion.android.nfc.NfcSetupActivity.onNewIntent(NfcSetupActivity.kt:101)
	at android.app.Activity.performNewIntent(Activity.java:8198)
	at android.app.Instrumentation.callActivityOnNewIntent(Instrumentation.java:1409)
	at android.app.Instrumentation.callActivityOnNewIntent(Instrumentation.java:1422)
	at android.app.ActivityThread.deliverNewIntents(ActivityThread.java:4001)
	at android.app.ActivityThread.handleNewIntent(ActivityThread.java:4013)
	at android.app.servertransaction.NewIntentItem.execute(NewIntentItem.java:53)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2277)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:246)
	at android.app.ActivityThread.main(ActivityThread.java:8425)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:596)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
	Caused by: java.io.IOException
	at android.nfc.tech.NdefFormatable.format(NdefFormatable.java:131)
	at android.nfc.tech.NdefFormatable.format(NdefFormatable.java:94)
	at io.homeassistant.companion.android.nfc.NFCUtil.writeMessageToTag(NFCUtil.kt:80)
	... 17 more

**Comment by dshokouhi** [comment]
@tefracky this error is different than what was reported here.  Can you please create a new issue and attach the full error please?  Please also mention what other NFC tags you have tried to use.  The error happens during formatting so we may need to add some debug logs to see what actually failed here.  The more descriptive the issue the better.

All I am going to close this issue now as the original issue was fixed by clearing data and also there is more than 1 error now in this ticket so we need to break things up.  If you still face an issue please create a new issue and get the crash logs as indicated up above so we can properly investigate this issue.

**Comment by bnjroos** [comment]
I had the same issue, I had to delete application data and restart the app and now it is working.
